packages/@eventual/aws-cdk/src/search/base-search-service.ts

```diff
@@ -0,0 +1,114 @@
+import { ENV_NAMES } from "@eventual/aws-runtime";
+import type { IndexSpec } from "@eventual/core/internal";
+import type { Function } from "aws-cdk-lib/aws-lambda";
+import { Runtime } from "aws-cdk-lib/aws-lambda";
+import * as aws_lambda_nodejs from "aws-cdk-lib/aws-lambda-nodejs";
+import { Lazy } from "aws-cdk-lib/core";
+import { Construct } from "constructs";
+import path from "path";
+import type { ServiceConstructProps } from "../service";
+import type { ServiceEntityProps } from "../utils";
+import { SearchIndex } from "./search-index";
+import type {
+  SearchPrincipal,
+  SearchService,
+  ServiceIndices,
+} from "./search-service";
+
+export type SearchIndexOverrides<Service> = ServiceEntityProps<
+  Service,
+  "SearchIndex",
+  Partial<Omit<IndexSpec, "index" | "mappings">>
+>;
+
+export interface BaseSearchServiceProps<Service> extends ServiceConstructProps {
+  indices?: SearchIndexOverrides<Service>;
+}
+
+/**
+ * Base Construct for the {@link SearchService} implementation in Eventual.
+ *
+ * It handles the creation of the Custom Resource for managing indices on
+ * the cluster and provisions all of the indices discovered in the Eventual
+ * application.
+ */
+export abstract class BaseSearchService<Service>
+  extends Construct
+  implements SearchService<Service>
+{
+  /**
+   * The OpenSearch cluster's HTTPS endpoint
+   */
+  public abstract readonly endpoint: string;
+  /**
+   * Lambda Function that handles all Custom Resource requests for managing the Cluster.
+   *
+   * e.g. creating Indices
+   */
+  public readonly customResourceHandler;
+  /**
+   * Key-value of all the indices
+   */
+  public readonly indices: ServiceIndices<Service>;
+  /**
+   * Construct that is the root of the tree containing all Index Custom Resources
+   */
+  private readonly indexRoot: Construct;
+
+  constructor(props: BaseSearchServiceProps<Service>) {
+    super(props.systemScope, "SearchService");
+
+    this.indexRoot = new Construct(props.serviceScope, "Indices");
+    this.customResourceHandler = new aws_lambda_nodejs.NodejsFunction(
+      this,
+      "CustomResourceHandler",
+      {
+        entry: path.join(__dirname, "custom-resource", "index.js"),
+        handler: "index.handle",
+        memorySize: 512,
+        runtime: Runtime.NODEJS_18_X,
+        environment: {
+          OS_ENDPOINT: Lazy.string({
+            produce: () => this.endpoint,
+          }),
+          NODE_OPTIONS: "--enable-source-maps",
+        },
+      }
+    );
+    this.indices = Object.fromEntries(
+      props.build.search.indices.map((index) => {
+        const overrides =
+          props.indices?.[index.index as keyof typeof props.indices];
+        const spec: IndexSpec = {
+          ...(overrides ?? {}),
+          ...index,
+          settings: {
+            ...index.settings,
+            ...(overrides?.settings ?? {}),
```

Comment:
Is the precedence here right? overrides and then index and then reversed? Should we be consistent with the logic in service function?

Reply:
Oops, that's wrong.